userspace/falco/falco.cpp

```diff
@@ -838,73 +517,30 @@ int falco_init(int argc, char **argv)
 		engine->add_source(syscall_source, syscall_filter_factory, syscall_formatter_factory);
 		engine->add_source(k8s_audit_source, k8s_audit_filter_factory, k8s_audit_formatter_factory);
 
-		if(disable_sources.size() > 0)
+		for(const auto &src : app.copts().disable_sources)
 		{
-			auto it = disable_sources.begin();
-			while(it != disable_sources.end())
-			{
-				if(*it != syscall_source && *it != k8s_audit_source)
-				{
-					it = disable_sources.erase(it);
-					continue;
-				}
-				++it;
-			}
-			disable_syscall = disable_sources.count(syscall_source) > 0;
-			disable_k8s_audit = disable_sources.count(k8s_audit_source) > 0;
-			if (disable_syscall && disable_k8s_audit) {
-				throw std::invalid_argument("The event source \"syscall\" and \"k8s_audit\" can not be disabled together");
-			}
+			enabled_sources.erase(src);
 		}
 
-		// Some combinations of arguments are not allowed.
-		if (daemon && pidfilename == "") {
-			throw std::invalid_argument("If -d is provided, a pid file must also be provided");
-		}
-
-		ifstream conf_stream;
-		if (conf_filename.size())
-		{
-			conf_stream.open(conf_filename);
-			if (!conf_stream.is_open())
-			{
-				throw std::runtime_error("Could not find configuration file at " + conf_filename);
-			}
-		}
-		else
+		// XXX/mstemm technically this isn't right, you could disable syscall *and* k8s_audit and configure a plugin.
```

Comment:
True. Maybe we might move this check after the plugin loading logic, and re-populate this array there if a source plugin is loaded. Besides, this is a reminder for ourselves that we might want to re-factor this logic once `k8s_audit` gets moved out of Falco to become a plugin. From that moment on, I think it would not make sense to have an `enabled_sources` array and a `disable source` option, as Falco would support one of syscalls and plugin source with mutual exclusion.

Reply:
Btw, I believe we should port the current `k8s_audit` impl. to a plugin __before__ merging this PR. 
That would solve several micro issues like the above one.

Reply:
Yeah, I thought a good time to revisit this would be when we make initialization an application object instead of just dumped in main. I don't think it's very likely that somebody would want to run falco with plugins and also disable all other sources.

I really don't agree about waiting for k8s_audit being a plugin first, though. There are a few other things we will need to figure out to move k8s_audit to a plugin:
1) Are we going to keep just the webhook-based approach, leaving most of the configuration/setup outside of falco? Or should we have separate k8s_audit plugins for distros like eks, gke, etc that have ways to read and transform proprietary logs into audit events (https://github.com/sysdiglabs/stackdriver-webhook-bridge, https://github.com/sysdiglabs/ekscloudwatch). There may even be a way to do this in a distro-independent way by hooking in at the admission controller.
2) We will need to figure out how to run falco with multiple sets of source plugins running at once, or with a combination of syscalls and plugin events at the same time. Remember that currently, if a source plugin is configured, other events are disabled.
3) We need to decide whether to update k8s_audit_rules.yaml to be more plugin friendly, or to strive for 100% compatibility with the existing rules file, even if we have to change the plugin framework. It *might* be the case that the json plugin can handle the jevt.xxx fields currently used in k8s_audit_rules.yaml, but we'd want to do some testing to make sure.
4) The event model and fields support used for k8s_audit events is very slightly different than sinsp events. For example, for jevt events, a field can actually return a list of items rather than a single item. For example, `ka.req.pod.containers.privileged` returns a list of boolean values for all containers in a pod. I don't believe the sinsp filtercheck model supports this right now, so we'll need to add it.

To be clear, I'm 100% for converting k8s audit support into a plugin so we can get rid of the parallel implementation of events in userspace/engine/json_evt.cpp. A plugin-based approach will be way cleaner and better. I'm just pointing out that it's a lot of work and I wouldn't want to hold this up to wait for it.

Reply:
I believe the k8s_audit plugin will be ready very soon. Moreover, I guess the changeset to clean up the old k8s implementation will be quicker to review than this PR.

Also, to reply to your points:

1. We will likely provide the webhook-based approach as a plugin, along with other plugins, one for each distro (eks, gke, etc..). @Issif, @jasondellaluce, and I have done some experiments, and we will propose something soon, I hope :smile_cat:  Moreover, I'm pretty sure the admission controller approach cannot work to replace the current functionality (since not all the information available in the audit log is also provided to the admission controller. One notable example is the API response object.
2. Totally agree with that. But I guess that's something we cannot fix soon, isn't it?
3. I would like to keep `k8s_audit_rules.yaml` as is now (at least, I would not introduce any backward-incompatible change). It seems to me that ~100% is possible, but I agree we need to do some testing.
4. I'm aware of that issue, and fixing that should be prioritized.


That being said, my advice to wait is not really a blocker, though I still think it would be better.

Reply:
I just opened a PR in libs trying to fix point #4 👉🏼 https://github.com/falcosecurity/libs/pull/200

I generally agree with @leogr, but in this case I would accept merging this before moving k8s-audit, and think about this discussion later.